qiskit/algorithms/linear_solvers/hhl.py

```diff
@@ -0,0 +1,357 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""The HHL algorithm."""
+
+from typing import Optional, Union, List, Callable, Tuple
+import numpy as np
+
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit import QuantumCircuit, QuantumRegister, AncillaRegister
+from qiskit.circuit.library import PhaseEstimation
+from qiskit.circuit.library.arithmetic.piecewise_chebyshev import PiecewiseChebyshev
+from qiskit.opflow import Z, I, StateFn, TensoredOp
+from .exact_inverse import ExactInverse
+from .linear_solver import LinearSolver, LinearSolverResult
+from .observables.linear_system_observable import LinearSystemObservable
+
+
+# logger = logging.getLogger(__name__)
+
+class HHL(LinearSolver):
+    """The HHL algorithm to solve systems of linear equations"""
+    def __init__(self, epsilon: float = 1e-2) -> None:
+        """
+        Args:
+         epsilon : Error tolerance.
+
+        .. note::
+
+        References:
+
+        [1]: Carrera Vazquez, A., Hiptmair, R., & Woerner, S. (2020).
+             Enhancing the Quantum Linear Systems Algorithm using Richardson Extrapolation.
+             `arXiv:2009.04484 <http://arxiv.org/abs/2009.04484>`_
+
+        """
+        super().__init__()
+
+        self._epsilon = epsilon
+        # Tolerance for the different parts of the algorithm as per [1]
+        self._epsilon_r = epsilon / 3  # conditioned rotation
+        self._epsilon_s = epsilon / 3  # state preparation
+        self._epsilon_a = epsilon / 6  # hamiltonian simulation
+
+        # Number of qubits for the different registers
+        self._nb = None  # number of qubits for the solution register
+        self._nl = None  # number of qubits for the eigenvalue register
+        self._na = None  # number of ancilla qubits
+        self._nf = None  # number of flag qubits
+        # Time of the evolution. Once the matrix is specified,
+        # it can be updated 2 * np.pi / lambda_max
+        self._evo_time = 2 * np.pi
+        self._lambda_max = None
+        self._lambda_min = None
+        self._kappa = 1  # Set default condition number to 1
+
+        # Circuits for the different blocks of the algorithm
+        self._vector_circuit = None
+        self._matrix_circuit = None
+        self._inverse_circuit = None
+        self._post_rotation = None
+
+        # For now the default inverse implementation is exact
+        self._exact_inverse = True
+
+    @property
+    def _num_qubits(self) -> int:
+        """Return the number of qubits needed in the circuit.
+
+        Returns:
+            The total number of qubits.
+        """
+        num_qubits = self._na + self._nl + self._nl + self._nf
+
+        return num_qubits
+
+    def get_delta(self, n_l: int, lambda_min: float, lambda_max: float) -> float:
```

Comment:
Would this be a better marked as a private method with leading _. I see it used from hhl below itself but is it useful to be called externally where its part of the supported public API of the algorithm?

Reply:
marked private together with `calculate_norm` and `calculate_observable`